feat: update okta login guide for realworld app (#4883)
cypress-io/cypress-documentation: content/guides/end-to-end-testing/okta-authentication.md

## <Icon name="graduation-cap"></Icon> What you'll learn

> - Login to [Okta](https://okta.com) through the UI with

Reviewer: ```suggestion
- Log in to [Okta](https://okta.com) through the UI with
```

Author: addressed in b94aadc43

> ## <Icon name="graduation-cap"></Icon> What you'll learn

Reviewer 2: pretty much the same feedback as https://github.com/cypress-io/cypress-documentation/pull/4882

Author: once we settle on #4882 ill add the same feedback here

Author: made same updates from #4882 in 31b0c6c52

## Custom Command for Okta Authentication

There are two ways you can authenticate to Okta

- [Login with cy.origin()](/guides/end-to-end-testing/okta-authentication#Login-with-cy-origin)
- [Legacy Programmatic Access](/guides/end-to-end-testing/okta-authentication#Programmatic-Login-Legacy)

### Login with [`cy.origin()`](/api/commands/origin)

> Next, we'll write a command to perform a login to [Okta](https://okta.com) using

Reviewer 2: ```suggestion
Next, we'll write a custom command to perform a login to [Okta](https://okta.com) using
```

Author: addressed in 31b0c6c52

## Custom Command for Okta Authentication

There are two ways you can authenticate to Okta:

> - [Login with cy.origin()](/guides/end-to-end-testing/okta-authentication#Login-with-cy-origin)

Author: addressed in b94aadc43

[Okta](https://okta.com). This command will use
[`cy.origin()`](/api/commands/origin) to

1. navigate to the Okta origin
2. input user credentials
3. sign in and redirect back to the
   [Cypress Real World App](https://github.com/cypress-io/cypress-realworld-app)
4. cache the results with [`cy.session()`](/api/commands/session)

```jsx
// cypress/support/auth-provider-commands/okta.ts
// Okta
const loginToOkta = (username: string, password: string) => {
  Cypress.log({
    displayName: 'OKTA LOGIN',
    message: [`🔐 Authenticating | ${username}`],
    autoEnd: false,
  })

  cy.visit('/')
  cy.origin(
    Cypress.env('okta_domain'),
    { args: { username, password } },
    ({ username, password }) => {
      cy.get('input[name="identifier"]').type(username)
      cy.get('input[name="credentials.passcode"]').type(password, {
        log: false,
      })
      cy.get('[type="submit"]').click()
    }
  )

  cy.get('[data-test="sidenav-username"]').should('contain', username)
}
// right now our custom command is light. More on this later!
Cypress.Commands.add('loginByOkta', (username: string, password: string) => {
  return loginToOkta(username, password)
})
```

> Secondly, we can use our `loginByOkta` command in the test. Below is our test to

Reviewer 2: ```suggestion
Now, we can use our `loginByOkta` command in the test. Below is our test to
```

Author: addressed in b94aadc43

> login as a user via [Okta](https://okta.com) and run a few basic sanity tests.

Reviewer 2: ```suggestion
login as a user via [Okta](https://okta.com) and run a few basic sanity checks.
```
only one test in the example

Author: addressed in b94aadc43

<Alert type="success">

The

> [runnable version of this test](https://github.com/cypress-io/cypress-realworld-app/blob/develop/cypress/tests/ui-auth-providers/okta.spec.ts)

Reviewer 2: Is the custom command found in this file or the support file?

Author: support file, which in the code snippet at the top has something like `cypress/support/auth-provider-commands/okta.ts`. I'm hopiong thats enough to make it discoverable since that is what we've done elsewhere, even though just clicking on the custom command just takes you the the typings definition

</Alert>

```jsx
describe('Okta', function () {
  beforeEach(function () {
    cy.task('db:seed')
    cy.loginByOkta(Cypress.env('okta_username'), Cypress.env('okta_password'))
  })

  it('verifies signed in user does not have a bank account', function () {
    cy.get('[data-test="sidenav-bankaccounts"]').click()
    cy.get('[data-test="empty-list-header"]').should('be.visible')
  })
})
```

<DocsVideo src="/img/examples/okta-origin.mp4"></DocsVideo>

Lastly, we can refactor our login command to take advantage of
[`cy.session()`](/api/commands/session) to store our logged in user so we don't
have to reauthenticate with everything test.

```jsx
Cypress.Commands.add('loginByOkta', (username: string, password: string) => {
  cy.session(
    `okta-${username}`,
    () => {
      return loginToOkta(username, password)
    },
    {
      validate() {
        cy.visit('/')
        cy.get('[data-test="sidenav-username"]').should('contain', username)
      },
    }
  )
})
```

<DocsVideo src="/img/examples/okta-session-restore.mp4"></DocsVideo>

### Programmatic Login

Next, we will write a command named `loginByOktaApi` to perform a programmatic

> login into [Okta](https://okta.com) and set an item in localStorage with the

Reviewer 2: ```suggestion
login into [Okta](https://okta.com) and set an item in `localStorage` with the
```

Author: addressed in b94aadc43

authenticated users details, which we will use in our application code to verify
we are authenticated under test.

In order to make sure this is enabled inside the
[Cypress Real World App](https://github.com/cypress-io/cypress-realworld-app),

> Please set the `REACT_APP_OKTA_PROGRAMMATIC` environment variable to `true`.

Reviewer 2: Why is this ENV needed? Also do we need to call it out here or can we link to a section in the read me to share how to set this up?

Author: I added `REACT_APP_OKTA_PROGRAMMATIC` since the programmatic approach strips the auth guards out of the webapp, which we don't want with UI login with `cy.origin()` since we want to test how the app actually behaves and get the natural redirect to the auth page when we aren't authenticated

Author: I figured here is fine since the README just points back to the guide

</Alert>

### Adapting an Okta App for Testing

<Alert type="info">

<strong class="alert-header">Note</strong>

> The previous sections focused on the programmatic Okta authentication practice

Reviewer 2: ```suggestion
The previous section focused on the programmatic Okta authentication practice
```

Author: addressed in b94aadc43

> within Cypress tests. To use this practice it is assumed you are testing an app

Reviewer 2: ```suggestion
within Cypress tests. To use this practice, it is assumed you are testing an app
```

Author: addressed in b94aadc43
